machine_learning/polynomial_regression.py

```diff
@@ -0,0 +1,216 @@
+"""
+Polynomial regression is a type of regression analysis that models the relationship
+between a predictor x and the response y as an mth-degree polynomial:
+
+y = β₀ + β₁x + β₂x² + ... + βₘxᵐ + εᵢ
+
+By treating x, x², ..., xᵐ as distinct variables, we see that polynomial regression is a
+special case of multiple linear regression. Therefore, we can use ordinary least squares
+(OLS) estimation to estimate the vector of model parameters β = (β₀, β₁, β₂, ..., βₘ)
+for polynomial regression:
+
+β = (XᵀX)⁻¹Xᵀy = X⁺y
+
+where X is the design matrix, y is the response vector, and X⁺ denotes the Moore–Penrose
+pseudoinverse of X. In the case of polynomial regression, the design matrix is
+
+    |1  x₁  x₁² ⋯ x₁ᵐ|
+X = |1  x₂  x₂² ⋯ x₂ᵐ|
+    |⋮  ⋮   ⋮   ⋱ ⋮  |
+    |1  xₙ  xₙ² ⋯  xₙᵐ|
+
+In OLS estimation, inverting XᵀX to compute X⁺ can be very numerically unstable. This
+implementation sidesteps this need to invert XᵀX by computing X⁺ using singular value
+decomposition (SVD):
+
+β = VΣ⁺Uᵀy
+
+where UΣVᵀ is an SVD of X.
+
+References:
+    - https://en.wikipedia.org/wiki/Polynomial_regression
+    - https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse
+    - https://en.wikipedia.org/wiki/Numerical_methods_for_linear_least_squares
+    - https://en.wikipedia.org/wiki/Singular_value_decomposition
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+class PolynomialRegression:
+    __slots__ = "degree", "params"
+
+    def __init__(self, degree: int) -> None:
+        """
+        @raises ValueError: if the polynomial degree is negative
+
+        >>> pass
+        """
+        if degree < 0:
+            raise ValueError("Polynomial degree must be non-negative")
+
+        self.degree = degree
+        self.params = None
+
+    @staticmethod
+    def _design_matrix(data: np.ndarray, degree: int) -> np.ndarray:
+        """
+        Constructs a polynomial regression design matrix for the given input data. For
+        input data x = (x₁, x₂, ..., xₙ) and polynomial degree m, the design matrix is
+        the Vandermonde matrix
+
+            |1  x₁  x₁² ⋯ x₁ᵐ|
+        X = |1  x₂  x₂² ⋯ x₂ᵐ|
+            |⋮  ⋮   ⋮   ⋱ ⋮  |
+            |1  xₙ  xₙ² ⋯  xₙᵐ|
+
+        Reference: https://en.wikipedia.org/wiki/Vandermonde_matrix
+
+        @param data:    the input predictor values x, either for model fitting or for
+                        prediction
+        @param degree:  the polynomial degree m
+        @returns:       the Vandermonde matrix X (see above)
+        @raises ValueError: if input data is not N x 1
+
+        >>> x = np.array([0, 1, 2])
+        >>> PolynomialRegression._design_matrix(x, degree=0)
+        array([[1],
+               [1],
+               [1]])
```

Comment:
```suggestion
        array([[1], [1], [1]])
```
These arrays don't have to be broken across so many lines

Reply:
Doctests look for exact matches, so I'm pretty sure removing the whitespace would break this doctest